build: allow to customize tmpdir

Kikobeats opened the pull request:
Hello,

This change allows customizing the directory for temporary files, setting up the current behavior as default.

I'm particularly interested in this change because I want to use a memory filesystem (`memfs`) to speed up a bit the puppeteer process.

Kikobeats committed:
build: allow to customize tmpdir

Kikobeats commented:
hum checks are failing for a non related thing.

Kikobeats committed:
Update Launcher.ts

mathiasbynens reviewed (changes requested):
The option is named `tmpdir` but it should be `tmpDir` to match the casing convention.

If we want to do this then introducing an [environment variable](https://github.com/puppeteer/puppeteer/blob/v9.1.1/docs/api.md#environment-variables) like `PUPPETEER_TMP_DIR` seems like a cleaner option.

The patch would also need to include a test and the relevant changes to documentation. WDYT?

Kikobeats committed:
Update Launcher.ts

Kikobeats commented:
ops, fixed the camel case naming 🙂

I'm not sure about `PUPPETEER_TMP_DIR` because it's limiting the possibility to use a function where the path can change over time, rather than a static string.

Kikobeats committed:
Merge branch 'main' into patch-2

mathiasbynens commented:
> I'm not sure about `PUPPETEER_TMP_DIR` because it's limiting the possibility to use a function where the path can change over time, rather than a static string.

If the path changes over time and your code knows about it, can't you have it update the `PUPPETEER_TMP_DIR` env var accordingly?

Kikobeats commented:
you're right; I'm assuming `os.tmpdir()` will return a different path every time, but it looks like constant value:

<img width="415" alt="Screen Shot 2021-05-17 at 11 06 05" src="https://user-images.githubusercontent.com/2096101/118463294-e633c400-b6ff-11eb-82e9-bba9e6d91e74.png">

let me do the code changes for a new environment variable 🙂

Kikobeats committed:
Merge branch 'main' into patch-2

Kikobeats committed:
refactor: add .tmpDir helper

Kikobeats commented:
@mathiasbynens done at https://github.com/puppeteer/puppeteer/pull/7243/commits/083ea5ff7e8f6a51b16db9699e237864961c9eaa

Do you think it could be moved into a common, so it just could be declared once?

Kikobeats committed:
docs: add PUPPETEER_TMP_DIR

Kikobeats committed:
Merge branch 'main' into patch-2

danpark-salesforce committed:
feature: add drag-and-drop support. (#7150)


This commit adds drag-and-drop support, leveraging new additions to the CDP
Input domain (Input.setInterceptDrags, Input.dispatchDragEvent, and
Input.dragIntercepted).

Kikobeats committed:
Merge branch 'main' into patch-2

Kikobeats commented:
@mathiasbynens Can I do something to move this PR forward? 🙂

mathiasbynens commented:
The diff seems to include lots of unrelated changes. Could you please check?

Kikobeats committed:
Merge branch 'main' into patch-2

Kikobeats commented:
@mathiasbynens updated 👍

mathiasbynens commented on src/node/Launcher.ts:
```diff
@@ -29,8 +29,11 @@ import {
+
 import { Product } from '../common/Product.js';
 
+const tmpDir = () => process.env.PUPPETEER_TMP_DIR || os.tmpdir()
```
Could you please fix the lint errors?

Kikobeats committed:
Update Launcher.js

Kikobeats committed:
Update Launcher.ts

Kikobeats committed:
Update Launcher.js

Kikobeats committed:
fix: linter

Kikobeats commented:
@mathiasbynens all green now 🙂

mathiasbynens commented on experimental/puppeteer-firefox/lib/Launcher.js:
```diff
@@ -30,7 +30,9 @@ const WebSocketTransport = require('./WebSocketTransport');
-const FIREFOX_PROFILE_PATH = path.join(os.tmpdir(), 'puppeteer_firefox_profile-');
+const tmpDir = () => process.env.PUPPETEER_TMP_DIR || os.tmpdir();
+
+const FIREFOX_PROFILE_PATH = path.join(tmpDir(),'puppeteer_firefox_profile-');
```
```suggestion
const FIREFOX_PROFILE_PATH = path.join(tmpDir(), 'puppeteer_firefox_profile-');
```

mathiasbynens commented on docs/api.md:
```diff
@@ -454,6 +454,7 @@ If Puppeteer doesn't find them in the environment during the installation step,
 
 - `HTTP_PROXY`, `HTTPS_PROXY`, `NO_PROXY` - defines HTTP proxy settings that are used to download and run Chromium.
 - `PUPPETEER_SKIP_CHROMIUM_DOWNLOAD` - do not download bundled Chromium during installation step.
+- `PUPPETEER_TMP_DIR` - defines the directory to be used by Puppeteer for creating temporary files. It's [os.tmpdir](https://nodejs.org/api/os.html#os_os_tmpdir) by default.
```
```suggestion
- `PUPPETEER_TMP_DIR` - defines the directory to be used by Puppeteer for creating temporary files. Defaults to [`os.tmpdir()`](https://nodejs.org/api/os.html#os_os_tmpdir).
```

Kikobeats committed:
Update experimental/puppeteer-firefox/lib/Launcher.js

Co-authored-by: Mathias Bynens <mathias@qiwi.be>

Kikobeats committed:
Update docs/api.md

Co-authored-by: Mathias Bynens <mathias@qiwi.be>

Kikobeats commented:
@mathiasbynens I think now can be merged

jschfflr committed:
Merge branch 'main' into patch-2

jschfflr reviewed (changes requested):
Could you maybe move the implementation of `const tmpDir = () => process.env.PUPPETEER_TMP_DIR || os.tmpdir();` into a separate file like `helpers.ts` or something similar to reduce code duplication?

Kikobeats committed:
Merge branch 'main' into patch-2

Kikobeats committed:
Merge branch 'main' into patch-2

Kikobeats commented:
> Could you maybe move the implementation of `const tmpDir = () => process.env.PUPPETEER_TMP_DIR || os.tmpdir();` into a separate file like `helpers.ts` or something similar to reduce code duplication?

I want to do that, `experimental/puppeteer-firefox/lib/Launcher.js` is CJS and `src/node/Launcher.ts` is ESM.

I guess I can't just create a file that can be loaded on both parts; am I wrong?

jschfflr commented:
> > Could you maybe move the implementation of `const tmpDir = () => process.env.PUPPETEER_TMP_DIR || os.tmpdir();` into a separate file like `helpers.ts` or something similar to reduce code duplication?
> 
> I want to do that, `experimental/puppeteer-firefox/lib/Launcher.js` is CJS and `src/node/Launcher.ts` is ESM.
> 
> I guess I can't just create a file that can be loaded on both parts; am I wrong?

Ah sorry, I didn't see that one is for Firefox and one in src. In this case, I think this change looks good!

Kikobeats committed:
Merge branch 'main' into patch-2

Kikobeats committed:
Merge branch 'main' into patch-2

Kikobeats commented:
@jschfflr @mathiasbynens any chance to merge this? it's already approved and tests pass on green 🙂

jschfflr committed:
Merge branch 'main' into patch-2
